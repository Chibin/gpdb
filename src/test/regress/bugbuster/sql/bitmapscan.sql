drop table if exists bug;
drop table if exists bug2;

create table bug as select '2008-02-01'::DATE AS DT, 
	case when j <= 96 
		then 0 
	when j<= 98 then 2 
	when j<= 99 then 3 
	when i % 1000 < 900 then 4 
	when i % 1000 < 800 then 5 
	when i % 1000 <= 998 then 5 else 6 
	end as ind, 
	(i*117-j)::bigint as s from generate_series(1,100) i, generate_series(1,100) j distributed randomly;

create table bug2 as select * from bug;

insert into bug select DT + 1, ind, s from bug2;
insert into bug select DT + 2, ind, s from bug2;
insert into bug select DT + 3, ind, s from bug2;
insert into bug select DT + 4, ind, s from bug2;
insert into bug select DT + 5, ind, s from bug2;
insert into bug select DT + 6, ind, s from bug2;
insert into bug select DT + 7, ind, s from bug2;
insert into bug select DT + 8, ind, s from bug2;
insert into bug select DT + 9, ind, s from bug2;
insert into bug select DT + 10, ind, s from bug2;
insert into bug select DT + 11, ind, s from bug2;
insert into bug select DT + 12, ind, s from bug2;
insert into bug select DT + 13, ind, s from bug2;
insert into bug select DT + 14, ind, s from bug2;
insert into bug select DT + 15, ind, s from bug2;

create index bug_idx on bug using bitmap (ind, dt);

vacuum analyze bug;
create table mpp4593_bmbug (dt date, ind int, s bigint);
create index bmbug_idx on mpp4593_bmbug using bitmap(ind, dt);
-- Test Bitmap Indexscan
drop table if exists test;
create table test (a integer, b integer) distributed by (a);
insert into test select a, a%25 from generate_series(1,100) a;
create index test_a on test (a);
set enable_seqscan=off;
set enable_indexscan=off;
set enable_bitmapscan=on;

-- start_ignore
select * from test where a<10;
-- end_ignore

-- returnning one or more tuples
select * from test where a<10;

-- start_ignore
select * from test where a>100;
-- end_ignore

-- returninig no tuples
select * from test where a>100;
-- Test Bitmap Heapscan + Bitmap OR + Bitmap Indexscan

drop table if exists test;
create table test (a integer, b integer) distributed by (a);
insert into test select a, a%25 from generate_series(1,100) a;

-- Test on 2 btrees
create index test_a on test (a);
create index test_b on test (b);

set enable_seqscan=off;
set enable_indexscan=off;
set enable_bitmapscan=on;

-- start_ignore
select * from test where a<100 or b>10;
-- end_ignore

select * from test where a<100 or b>10;

-- start_ignore
select * from test where a<100 or b>30;
-- end_ignore

-- Returnning no tuples from one branch
select * from test where a<100 or b>30;

-- start_ignore
select * from test where a<1 or b>30;
-- end_ignore

-- Returnning no tuples from both branch
select * from test where a<1 or b>30;

-- Test on 2 bitmaps
drop index test_a;
drop index test_b;
create index test_bm_a on test using bitmap(a);
create index test_bm_b on test using bitmap(b);

-- start_ignore
select * from test where a<100 or b>10;
-- end_ignore

select * from test where a<100 or b>10;

-- start_ignore
select * from test where a<100 or b>30;
-- end_ignore

-- Returnning no tuples from one branch
select * from test where a<100 or b>30;

-- start_ignore
select * from test where a<1 or b>30;
-- end_ignore

-- Returnning no tuples from both branch
select * from test where a<1 or b>30;

-- Test on 1 btree, 1 bitmap
drop index test_bm_a;
drop index test_bm_b;
create index test_a on test (a);
create index test_bm_b on test using bitmap(b);

-- start_ignore
select * from test where a<100 or b>10;
-- end_ignore

select * from test where a<100 or b>10;

-- start_ignore
select * from test where a<100 or b>30;
-- end_ignore

-- Returnning no tuples from one branch
select * from test where a<100 or b>30;

-- start_ignore
select * from test where a<1 or b>30;
-- end_ignore

-- Returnning no tuples from both branch
select * from test where a<1 or b>30;
-- Test ArrayKeys
drop table if exists test;
create table test (a integer, b integer) distributed by (a);
insert into test select a, a%25 from generate_series(1,100) a;
create index test_a on test (a);
set enable_seqscan=off;
set enable_indexscan=off;
set enable_bitmapscan=on;

-- start_ignore
select * from test where a in (1,3,5);
-- end_ignore

select * from test where a in (1,3,5);
