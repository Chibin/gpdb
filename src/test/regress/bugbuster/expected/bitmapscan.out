drop table if exists bug;
NOTICE:  table "bug" does not exist, skipping
drop table if exists bug2;
NOTICE:  table "bug2" does not exist, skipping
create table bug as select '2008-02-01'::DATE AS DT, 
	case when j <= 96 
		then 0 
	when j<= 98 then 2 
	when j<= 99 then 3 
	when i % 1000 < 900 then 4 
	when i % 1000 < 800 then 5 
	when i % 1000 <= 998 then 5 else 6 
	end as ind, 
	(i*117-j)::bigint as s from generate_series(1,100) i, generate_series(1,100) j distributed randomly;
create table bug2 as select * from bug;
NOTICE:  Table doesn't have 'DISTRIBUTED BY' clause -- Using column(s) named 'dt' as the Greenplum Database data distribution key for this table.
HINT:  The 'DISTRIBUTED BY' clause determines the distribution of data. Make sure column(s) chosen are the optimal data distribution key to minimize skew.
insert into bug select DT + 1, ind, s from bug2;
insert into bug select DT + 2, ind, s from bug2;
insert into bug select DT + 3, ind, s from bug2;
insert into bug select DT + 4, ind, s from bug2;
insert into bug select DT + 5, ind, s from bug2;
insert into bug select DT + 6, ind, s from bug2;
insert into bug select DT + 7, ind, s from bug2;
insert into bug select DT + 8, ind, s from bug2;
insert into bug select DT + 9, ind, s from bug2;
insert into bug select DT + 10, ind, s from bug2;
insert into bug select DT + 11, ind, s from bug2;
insert into bug select DT + 12, ind, s from bug2;
insert into bug select DT + 13, ind, s from bug2;
insert into bug select DT + 14, ind, s from bug2;
insert into bug select DT + 15, ind, s from bug2;
create index bug_idx on bug using bitmap (ind, dt);
vacuum analyze bug;
create table mpp4593_bmbug (dt date, ind int, s bigint);
NOTICE:  Table doesn't have 'DISTRIBUTED BY' clause -- Using column named 'dt' as the Greenplum Database data distribution key for this table.
HINT:  The 'DISTRIBUTED BY' clause determines the distribution of data. Make sure column(s) chosen are the optimal data distribution key to minimize skew.
create index bmbug_idx on mpp4593_bmbug using bitmap(ind, dt);
-- Test Bitmap Indexscan
drop table if exists test;
NOTICE:  table "test" does not exist, skipping
create table test (a integer, b integer) distributed by (a);
insert into test select a, a%25 from generate_series(1,100) a;
create index test_a on test (a);
set enable_seqscan=off;
set enable_indexscan=off;
set enable_bitmapscan=on;
-- start_ignore
select * from test where a<10;
 a | b 
---+---
 1 | 1
 2 | 2
 8 | 8
 9 | 9
 3 | 3
 4 | 4
 5 | 5
 6 | 6
 7 | 7
(9 rows)

-- end_ignore
-- returnning one or more tuples
select * from test where a<10;
 a | b 
---+---
 1 | 1
 2 | 2
 3 | 3
 4 | 4
 5 | 5
 6 | 6
 7 | 7
 8 | 8
 9 | 9
(9 rows)

-- start_ignore
select * from test where a>100;
 a | b 
---+---
(0 rows)

-- end_ignore
-- returninig no tuples
select * from test where a>100;
 a | b 
---+---
(0 rows)

-- Test Bitmap Heapscan + Bitmap OR + Bitmap Indexscan
drop table if exists test;
create table test (a integer, b integer) distributed by (a);
insert into test select a, a%25 from generate_series(1,100) a;
-- Test on 2 btrees
create index test_a on test (a);
create index test_b on test (b);
set enable_seqscan=off;
set enable_indexscan=off;
set enable_bitmapscan=on;
-- start_ignore
select * from test where a<100 or b>10;
 a  | b  
----+----
  8 |  8
  9 |  9
 10 | 10
 11 | 11
 12 | 12
 23 | 23
 24 | 24
 25 |  0
 26 |  1
 27 |  2
 32 |  7
 43 | 18
 44 | 19
 45 | 20
 46 | 21
 47 | 22
 58 |  8
 59 |  9
 60 | 10
 61 | 11
 62 | 12
 64 | 14
 65 | 15
 66 | 16
 67 | 17
 78 |  3
 79 |  4
 80 |  5
 81 |  6
 82 |  7
 92 | 17
 93 | 18
 94 | 19
 95 | 20
 98 | 23
 99 | 24
  3 |  3
  4 |  4
  5 |  5
  6 |  6
  7 |  7
 18 | 18
 19 | 19
 20 | 20
 21 | 21
 22 | 22
 38 | 13
 39 | 14
 40 | 15
 41 | 16
 42 | 17
 53 |  3
 54 |  4
 55 |  5
 56 |  6
 57 |  7
 73 | 23
 74 | 24
 75 |  0
 76 |  1
 77 |  2
 87 | 12
 88 | 13
 89 | 14
 90 | 15
 91 | 16
 96 | 21
 97 | 22
  1 |  1
  2 |  2
 13 | 13
 14 | 14
 15 | 15
 16 | 16
 17 | 17
 28 |  3
 29 |  4
 30 |  5
 31 |  6
 33 |  8
 34 |  9
 35 | 10
 36 | 11
 37 | 12
 48 | 23
 49 | 24
 50 |  0
 51 |  1
 52 |  2
 63 | 13
 68 | 18
 69 | 19
 70 | 20
 71 | 21
 72 | 22
 83 |  8
 84 |  9
 85 | 10
 86 | 11
(99 rows)

-- end_ignore
select * from test where a<100 or b>10;
 a  | b  
----+----
  3 |  3
  4 |  4
  5 |  5
  6 |  6
  7 |  7
 18 | 18
 19 | 19
 20 | 20
 21 | 21
 22 | 22
 38 | 13
 39 | 14
 40 | 15
 41 | 16
 42 | 17
 53 |  3
 54 |  4
 55 |  5
 56 |  6
 57 |  7
 73 | 23
 74 | 24
 75 |  0
 76 |  1
 77 |  2
 87 | 12
 88 | 13
 89 | 14
 90 | 15
 91 | 16
 96 | 21
 97 | 22
  8 |  8
  9 |  9
 10 | 10
 11 | 11
 12 | 12
 23 | 23
 24 | 24
 25 |  0
 26 |  1
 27 |  2
 32 |  7
 43 | 18
 44 | 19
 45 | 20
 46 | 21
 47 | 22
 58 |  8
 59 |  9
 60 | 10
 61 | 11
 62 | 12
 64 | 14
 65 | 15
 66 | 16
 67 | 17
 78 |  3
 79 |  4
 80 |  5
 81 |  6
 82 |  7
 92 | 17
 93 | 18
 94 | 19
 95 | 20
 98 | 23
 99 | 24
  1 |  1
  2 |  2
 13 | 13
 14 | 14
 15 | 15
 16 | 16
 17 | 17
 28 |  3
 29 |  4
 30 |  5
 31 |  6
 33 |  8
 34 |  9
 35 | 10
 36 | 11
 37 | 12
 48 | 23
 49 | 24
 50 |  0
 51 |  1
 52 |  2
 63 | 13
 68 | 18
 69 | 19
 70 | 20
 71 | 21
 72 | 22
 83 |  8
 84 |  9
 85 | 10
 86 | 11
(99 rows)

-- start_ignore
select * from test where a<100 or b>30;
 a  | b  
----+----
  8 |  8
  9 |  9
 10 | 10
 11 | 11
 12 | 12
 23 | 23
 24 | 24
 25 |  0
 26 |  1
 27 |  2
 32 |  7
 43 | 18
 44 | 19
 45 | 20
 46 | 21
 47 | 22
 58 |  8
 59 |  9
 60 | 10
 61 | 11
 62 | 12
 64 | 14
 65 | 15
 66 | 16
 67 | 17
 78 |  3
 79 |  4
 80 |  5
 81 |  6
 82 |  7
 92 | 17
 93 | 18
 94 | 19
 95 | 20
 98 | 23
 99 | 24
  1 |  1
  2 |  2
 13 | 13
 14 | 14
 15 | 15
 16 | 16
 17 | 17
 28 |  3
 29 |  4
 30 |  5
 31 |  6
 33 |  8
 34 |  9
 35 | 10
 36 | 11
 37 | 12
 48 | 23
 49 | 24
 50 |  0
 51 |  1
 52 |  2
 63 | 13
 68 | 18
 69 | 19
 70 | 20
 71 | 21
 72 | 22
 83 |  8
 84 |  9
 85 | 10
 86 | 11
  3 |  3
  4 |  4
  5 |  5
  6 |  6
  7 |  7
 18 | 18
 19 | 19
 20 | 20
 21 | 21
 22 | 22
 38 | 13
 39 | 14
 40 | 15
 41 | 16
 42 | 17
 53 |  3
 54 |  4
 55 |  5
 56 |  6
 57 |  7
 73 | 23
 74 | 24
 75 |  0
 76 |  1
 77 |  2
 87 | 12
 88 | 13
 89 | 14
 90 | 15
 91 | 16
 96 | 21
 97 | 22
(99 rows)

-- end_ignore
-- Returnning no tuples from one branch
select * from test where a<100 or b>30;
 a  | b  
----+----
  8 |  8
  9 |  9
 10 | 10
 11 | 11
 12 | 12
 23 | 23
 24 | 24
 25 |  0
 26 |  1
 27 |  2
 32 |  7
 43 | 18
 44 | 19
 45 | 20
 46 | 21
 47 | 22
 58 |  8
 59 |  9
 60 | 10
 61 | 11
 62 | 12
 64 | 14
 65 | 15
 66 | 16
 67 | 17
 78 |  3
 79 |  4
 80 |  5
 81 |  6
 82 |  7
 92 | 17
 93 | 18
 94 | 19
 95 | 20
 98 | 23
 99 | 24
  1 |  1
  2 |  2
 13 | 13
 14 | 14
 15 | 15
 16 | 16
 17 | 17
 28 |  3
 29 |  4
 30 |  5
 31 |  6
 33 |  8
 34 |  9
 35 | 10
 36 | 11
 37 | 12
 48 | 23
 49 | 24
 50 |  0
 51 |  1
 52 |  2
 63 | 13
 68 | 18
 69 | 19
 70 | 20
 71 | 21
 72 | 22
 83 |  8
 84 |  9
 85 | 10
 86 | 11
  3 |  3
  4 |  4
  5 |  5
  6 |  6
  7 |  7
 18 | 18
 19 | 19
 20 | 20
 21 | 21
 22 | 22
 38 | 13
 39 | 14
 40 | 15
 41 | 16
 42 | 17
 53 |  3
 54 |  4
 55 |  5
 56 |  6
 57 |  7
 73 | 23
 74 | 24
 75 |  0
 76 |  1
 77 |  2
 87 | 12
 88 | 13
 89 | 14
 90 | 15
 91 | 16
 96 | 21
 97 | 22
(99 rows)

-- start_ignore
select * from test where a<1 or b>30;
 a | b 
---+---
(0 rows)

-- end_ignore
-- Returnning no tuples from both branch
select * from test where a<1 or b>30;
 a | b 
---+---
(0 rows)

-- Test on 2 bitmaps
drop index test_a;
drop index test_b;
create index test_bm_a on test using bitmap(a);
create index test_bm_b on test using bitmap(b);
-- start_ignore
select * from test where a<100 or b>10;
 a  | b  
----+----
  1 |  1
  2 |  2
 13 | 13
 14 | 14
 15 | 15
 16 | 16
 17 | 17
 28 |  3
 29 |  4
 30 |  5
 31 |  6
 33 |  8
 34 |  9
 35 | 10
 36 | 11
 37 | 12
 48 | 23
 49 | 24
 50 |  0
 51 |  1
 52 |  2
 63 | 13
 68 | 18
 69 | 19
 70 | 20
 71 | 21
 72 | 22
 83 |  8
 84 |  9
 85 | 10
 86 | 11
  8 |  8
  9 |  9
 10 | 10
 11 | 11
 12 | 12
 23 | 23
 24 | 24
 25 |  0
 26 |  1
 27 |  2
 32 |  7
 43 | 18
 44 | 19
 45 | 20
 46 | 21
 47 | 22
 58 |  8
 59 |  9
 60 | 10
 61 | 11
 62 | 12
 64 | 14
 65 | 15
 66 | 16
 67 | 17
 78 |  3
 79 |  4
 80 |  5
 81 |  6
 82 |  7
 92 | 17
 93 | 18
 94 | 19
 95 | 20
 98 | 23
 99 | 24
  3 |  3
  4 |  4
  5 |  5
  6 |  6
  7 |  7
 18 | 18
 19 | 19
 20 | 20
 21 | 21
 22 | 22
 38 | 13
 39 | 14
 40 | 15
 41 | 16
 42 | 17
 53 |  3
 54 |  4
 55 |  5
 56 |  6
 57 |  7
 73 | 23
 74 | 24
 75 |  0
 76 |  1
 77 |  2
 87 | 12
 88 | 13
 89 | 14
 90 | 15
 91 | 16
 96 | 21
 97 | 22
(99 rows)

-- end_ignore
select * from test where a<100 or b>10;
 a  | b  
----+----
  3 |  3
  4 |  4
  5 |  5
  6 |  6
  7 |  7
 18 | 18
 19 | 19
 20 | 20
 21 | 21
 22 | 22
 38 | 13
 39 | 14
 40 | 15
 41 | 16
 42 | 17
 53 |  3
 54 |  4
 55 |  5
 56 |  6
 57 |  7
 73 | 23
 74 | 24
 75 |  0
 76 |  1
 77 |  2
 87 | 12
 88 | 13
 89 | 14
 90 | 15
 91 | 16
 96 | 21
 97 | 22
  8 |  8
  9 |  9
 10 | 10
 11 | 11
 12 | 12
 23 | 23
 24 | 24
 25 |  0
 26 |  1
 27 |  2
 32 |  7
 43 | 18
 44 | 19
 45 | 20
 46 | 21
 47 | 22
 58 |  8
 59 |  9
 60 | 10
 61 | 11
 62 | 12
 64 | 14
 65 | 15
 66 | 16
 67 | 17
 78 |  3
 79 |  4
 80 |  5
 81 |  6
 82 |  7
 92 | 17
 93 | 18
 94 | 19
 95 | 20
 98 | 23
 99 | 24
  1 |  1
  2 |  2
 13 | 13
 14 | 14
 15 | 15
 16 | 16
 17 | 17
 28 |  3
 29 |  4
 30 |  5
 31 |  6
 33 |  8
 34 |  9
 35 | 10
 36 | 11
 37 | 12
 48 | 23
 49 | 24
 50 |  0
 51 |  1
 52 |  2
 63 | 13
 68 | 18
 69 | 19
 70 | 20
 71 | 21
 72 | 22
 83 |  8
 84 |  9
 85 | 10
 86 | 11
(99 rows)

-- start_ignore
select * from test where a<100 or b>30;
 a  | b  
----+----
  1 |  1
  2 |  2
 13 | 13
 14 | 14
 15 | 15
 16 | 16
 17 | 17
 28 |  3
 29 |  4
 30 |  5
 31 |  6
 33 |  8
 34 |  9
 35 | 10
 36 | 11
 37 | 12
 48 | 23
 49 | 24
 50 |  0
 51 |  1
 52 |  2
 63 | 13
 68 | 18
 69 | 19
 70 | 20
 71 | 21
 72 | 22
 83 |  8
 84 |  9
 85 | 10
 86 | 11
  3 |  3
  4 |  4
  5 |  5
  6 |  6
  7 |  7
 18 | 18
 19 | 19
 20 | 20
 21 | 21
 22 | 22
 38 | 13
 39 | 14
 40 | 15
 41 | 16
 42 | 17
 53 |  3
 54 |  4
 55 |  5
 56 |  6
 57 |  7
 73 | 23
 74 | 24
 75 |  0
 76 |  1
 77 |  2
 87 | 12
 88 | 13
 89 | 14
 90 | 15
 91 | 16
 96 | 21
 97 | 22
  8 |  8
  9 |  9
 10 | 10
 11 | 11
 12 | 12
 23 | 23
 24 | 24
 25 |  0
 26 |  1
 27 |  2
 32 |  7
 43 | 18
 44 | 19
 45 | 20
 46 | 21
 47 | 22
 58 |  8
 59 |  9
 60 | 10
 61 | 11
 62 | 12
 64 | 14
 65 | 15
 66 | 16
 67 | 17
 78 |  3
 79 |  4
 80 |  5
 81 |  6
 82 |  7
 92 | 17
 93 | 18
 94 | 19
 95 | 20
 98 | 23
 99 | 24
(99 rows)

-- end_ignore
-- Returnning no tuples from one branch
select * from test where a<100 or b>30;
 a  | b  
----+----
  1 |  1
  2 |  2
 13 | 13
 14 | 14
 15 | 15
 16 | 16
 17 | 17
 28 |  3
 29 |  4
 30 |  5
 31 |  6
 33 |  8
 34 |  9
 35 | 10
 36 | 11
 37 | 12
 48 | 23
 49 | 24
 50 |  0
 51 |  1
 52 |  2
 63 | 13
 68 | 18
 69 | 19
 70 | 20
 71 | 21
 72 | 22
 83 |  8
 84 |  9
 85 | 10
 86 | 11
  8 |  8
  9 |  9
 10 | 10
 11 | 11
 12 | 12
 23 | 23
 24 | 24
 25 |  0
 26 |  1
 27 |  2
 32 |  7
 43 | 18
 44 | 19
 45 | 20
 46 | 21
 47 | 22
 58 |  8
 59 |  9
 60 | 10
 61 | 11
 62 | 12
 64 | 14
 65 | 15
 66 | 16
 67 | 17
 78 |  3
 79 |  4
 80 |  5
 81 |  6
 82 |  7
 92 | 17
 93 | 18
 94 | 19
 95 | 20
 98 | 23
 99 | 24
  3 |  3
  4 |  4
  5 |  5
  6 |  6
  7 |  7
 18 | 18
 19 | 19
 20 | 20
 21 | 21
 22 | 22
 38 | 13
 39 | 14
 40 | 15
 41 | 16
 42 | 17
 53 |  3
 54 |  4
 55 |  5
 56 |  6
 57 |  7
 73 | 23
 74 | 24
 75 |  0
 76 |  1
 77 |  2
 87 | 12
 88 | 13
 89 | 14
 90 | 15
 91 | 16
 96 | 21
 97 | 22
(99 rows)

-- start_ignore
select * from test where a<1 or b>30;
 a | b 
---+---
(0 rows)

-- end_ignore
-- Returnning no tuples from both branch
select * from test where a<1 or b>30;
 a | b 
---+---
(0 rows)

-- Test on 1 btree, 1 bitmap
drop index test_bm_a;
drop index test_bm_b;
create index test_a on test (a);
create index test_bm_b on test using bitmap(b);
-- start_ignore
select * from test where a<100 or b>10;
 a  | b  
----+----
  1 |  1
  2 |  2
 13 | 13
 14 | 14
 15 | 15
 16 | 16
 17 | 17
 28 |  3
 29 |  4
 30 |  5
 31 |  6
 33 |  8
 34 |  9
 35 | 10
 36 | 11
 37 | 12
 48 | 23
 49 | 24
 50 |  0
 51 |  1
 52 |  2
 63 | 13
 68 | 18
 69 | 19
 70 | 20
 71 | 21
 72 | 22
 83 |  8
 84 |  9
 85 | 10
 86 | 11
  8 |  8
  9 |  9
 10 | 10
 11 | 11
 12 | 12
 23 | 23
 24 | 24
 25 |  0
 26 |  1
 27 |  2
 32 |  7
 43 | 18
 44 | 19
 45 | 20
 46 | 21
 47 | 22
 58 |  8
 59 |  9
 60 | 10
 61 | 11
 62 | 12
 64 | 14
 65 | 15
 66 | 16
 67 | 17
 78 |  3
 79 |  4
 80 |  5
 81 |  6
 82 |  7
 92 | 17
 93 | 18
 94 | 19
 95 | 20
 98 | 23
 99 | 24
  3 |  3
  4 |  4
  5 |  5
  6 |  6
  7 |  7
 18 | 18
 19 | 19
 20 | 20
 21 | 21
 22 | 22
 38 | 13
 39 | 14
 40 | 15
 41 | 16
 42 | 17
 53 |  3
 54 |  4
 55 |  5
 56 |  6
 57 |  7
 73 | 23
 74 | 24
 75 |  0
 76 |  1
 77 |  2
 87 | 12
 88 | 13
 89 | 14
 90 | 15
 91 | 16
 96 | 21
 97 | 22
(99 rows)

-- end_ignore
select * from test where a<100 or b>10;
 a  | b  
----+----
  8 |  8
  9 |  9
 10 | 10
 11 | 11
 12 | 12
 23 | 23
 24 | 24
 25 |  0
 26 |  1
 27 |  2
 32 |  7
 43 | 18
 44 | 19
 45 | 20
 46 | 21
 47 | 22
 58 |  8
 59 |  9
 60 | 10
 61 | 11
 62 | 12
 64 | 14
 65 | 15
 66 | 16
 67 | 17
 78 |  3
 79 |  4
 80 |  5
 81 |  6
 82 |  7
 92 | 17
 93 | 18
 94 | 19
 95 | 20
 98 | 23
 99 | 24
  3 |  3
  4 |  4
  5 |  5
  6 |  6
  7 |  7
 18 | 18
 19 | 19
 20 | 20
 21 | 21
 22 | 22
 38 | 13
 39 | 14
 40 | 15
 41 | 16
 42 | 17
 53 |  3
 54 |  4
 55 |  5
 56 |  6
 57 |  7
 73 | 23
 74 | 24
 75 |  0
 76 |  1
 77 |  2
 87 | 12
 88 | 13
 89 | 14
 90 | 15
 91 | 16
 96 | 21
 97 | 22
  1 |  1
  2 |  2
 13 | 13
 14 | 14
 15 | 15
 16 | 16
 17 | 17
 28 |  3
 29 |  4
 30 |  5
 31 |  6
 33 |  8
 34 |  9
 35 | 10
 36 | 11
 37 | 12
 48 | 23
 49 | 24
 50 |  0
 51 |  1
 52 |  2
 63 | 13
 68 | 18
 69 | 19
 70 | 20
 71 | 21
 72 | 22
 83 |  8
 84 |  9
 85 | 10
 86 | 11
(99 rows)

-- start_ignore
select * from test where a<100 or b>30;
 a  | b  
----+----
  3 |  3
  4 |  4
  5 |  5
  6 |  6
  7 |  7
 18 | 18
 19 | 19
 20 | 20
 21 | 21
 22 | 22
 38 | 13
 39 | 14
 40 | 15
 41 | 16
 42 | 17
 53 |  3
 54 |  4
 55 |  5
 56 |  6
 57 |  7
 73 | 23
 74 | 24
 75 |  0
 76 |  1
 77 |  2
 87 | 12
 88 | 13
 89 | 14
 90 | 15
 91 | 16
 96 | 21
 97 | 22
  8 |  8
  9 |  9
 10 | 10
 11 | 11
 12 | 12
 23 | 23
 24 | 24
 25 |  0
 26 |  1
 27 |  2
 32 |  7
 43 | 18
 44 | 19
 45 | 20
 46 | 21
 47 | 22
 58 |  8
 59 |  9
 60 | 10
 61 | 11
 62 | 12
 64 | 14
 65 | 15
 66 | 16
 67 | 17
 78 |  3
 79 |  4
 80 |  5
 81 |  6
 82 |  7
 92 | 17
 93 | 18
 94 | 19
 95 | 20
 98 | 23
 99 | 24
  1 |  1
  2 |  2
 13 | 13
 14 | 14
 15 | 15
 16 | 16
 17 | 17
 28 |  3
 29 |  4
 30 |  5
 31 |  6
 33 |  8
 34 |  9
 35 | 10
 36 | 11
 37 | 12
 48 | 23
 49 | 24
 50 |  0
 51 |  1
 52 |  2
 63 | 13
 68 | 18
 69 | 19
 70 | 20
 71 | 21
 72 | 22
 83 |  8
 84 |  9
 85 | 10
 86 | 11
(99 rows)

-- end_ignore
-- Returnning no tuples from one branch
select * from test where a<100 or b>30;
 a  | b  
----+----
  3 |  3
  4 |  4
  5 |  5
  6 |  6
  7 |  7
 18 | 18
 19 | 19
 20 | 20
 21 | 21
 22 | 22
 38 | 13
 39 | 14
 40 | 15
 41 | 16
 42 | 17
 53 |  3
 54 |  4
 55 |  5
 56 |  6
 57 |  7
 73 | 23
 74 | 24
 75 |  0
 76 |  1
 77 |  2
 87 | 12
 88 | 13
 89 | 14
 90 | 15
 91 | 16
 96 | 21
 97 | 22
  8 |  8
  9 |  9
 10 | 10
 11 | 11
 12 | 12
 23 | 23
 24 | 24
 25 |  0
 26 |  1
 27 |  2
 32 |  7
 43 | 18
 44 | 19
 45 | 20
 46 | 21
 47 | 22
 58 |  8
 59 |  9
 60 | 10
 61 | 11
 62 | 12
 64 | 14
 65 | 15
 66 | 16
 67 | 17
 78 |  3
 79 |  4
 80 |  5
 81 |  6
 82 |  7
 92 | 17
 93 | 18
 94 | 19
 95 | 20
 98 | 23
 99 | 24
  1 |  1
  2 |  2
 13 | 13
 14 | 14
 15 | 15
 16 | 16
 17 | 17
 28 |  3
 29 |  4
 30 |  5
 31 |  6
 33 |  8
 34 |  9
 35 | 10
 36 | 11
 37 | 12
 48 | 23
 49 | 24
 50 |  0
 51 |  1
 52 |  2
 63 | 13
 68 | 18
 69 | 19
 70 | 20
 71 | 21
 72 | 22
 83 |  8
 84 |  9
 85 | 10
 86 | 11
(99 rows)

-- start_ignore
select * from test where a<1 or b>30;
 a | b 
---+---
(0 rows)

-- end_ignore
-- Returnning no tuples from both branch
select * from test where a<1 or b>30;
 a | b 
---+---
(0 rows)

-- Test ArrayKeys
drop table if exists test;
create table test (a integer, b integer) distributed by (a);
insert into test select a, a%25 from generate_series(1,100) a;
create index test_a on test (a);
set enable_seqscan=off;
set enable_indexscan=off;
set enable_bitmapscan=on;
-- start_ignore
select * from test where a in (1,3,5);
 a | b 
---+---
 1 | 1
 3 | 3
 5 | 5
(3 rows)

-- end_ignore
select * from test where a in (1,3,5);
 a | b 
---+---
 1 | 1
 3 | 3
 5 | 5
(3 rows)

